### PR TITLE
Add possibility to modify the main.tf file by automatically removing modules and his references

### PR DESCRIPTION
**Pull request by maximenoel8**
This code is designed to remove clients module from the Terraform configuration file (`main.tf`) depending on a minion list declaration.
It adds three new arguments: 
 -  `use-tf-resource-cleaner` : will enable or disable the new feature to be compatible with previous version 
 -  `tf-resources-to-keep` : use for BV, declare the minions  you want to keep. The minions in https://github.com/SUSE/susemanager-ci/blob/9b5de70b0c57ec609aac8c8192b90ed53b487696/jenkins_pipelines/environments/manager-4.3-qe-build-validation-NUE#L4-L23 but not in `tf-resources-to-keep` list are going to be removed from the main.tf
By default, if `use-tf-resource-cleaner` is set and `tf-resources-to-keep` is empty ( or not declared), it will delete all the minions.
 -  `tf-resources-to-delete` : active action to delete proxy, monitoring-server and retail clients ( build and terminal clients ). If empty, those resources are kept, if declare, this resource is deleted. Deleting proxy without deleting the clients is probably a bad idea and need testing. We could have default behavior where if you remove proxy, you also remove the clients.

The modules by default excluded from this feature and that can't be deleted are : 
 - base
 - server
 - controller

 Here's a summary of what the code does:

1. **Extract Default Modules** (`get_default_modules`):
   - Identifies the modules in the `main.tf` file that should not be removed, specifically filtering out any modules related to "minion" or "client".

2. **Check if a Configuration Line References a Minion to Keep** (`contains_minion`):
   - Checks if a given line in the configuration references a minion that is on the list of minions to be retained.

3. **Filter Configuration Lines in the Controller Module** (`filter_module_references`):
   - Filters the lines in the controller module to ensure that only configurations related to the minions to keep are retained.
   - Ensures that the output module is not deleted and removes any lines marked as workarounds for clarity.

4. **Remove Unused Minion Modules** (`remove_unused_minion`):
   - Reads the `main.tf` file and identifies all the modules.
   - Extends the list of minions to keep by adding the default modules.
   - Iterates through the modules:
     - Removes any minion modules that are not in the list of minions to keep and are not part of the "base" configuration.
     - For the "controller" module, it applies the filtering to keep only the necessary configurations.
   - Cleans up any excessive newlines in the file.
   - Writes the cleaned-up content back to the `main.tf` file, logging which modules were removed.

### CODE REVIEW

**Comment by maximenoel8** [comment]
The naming is open to discussion, I'm never really good at finding explicit name.

**Comment by srbarrios** [comment]
> build and terminal minions

What's the reason to exclude terminal clients? Previous content written in their disk resource could not affect to the validation?

**Comment by srbarrios** [comment]
> proxy

At least on 4.3 MUs, we were also including shared client tools as part of the proxy onboarding, I couldn't confirm if this still the case. Here an [example](https://github.com/SUSE/spacewalk/issues/18394#issuecomment-1214848957).

We were doing it because we were first onboarding the proxy vm as a regular SLES15SP4 Salt Minion, and only after, we were setting it us a Proxy.

If that still the case, at least for 4.3, then we might also need to delete the proxy resources and re-onboard them. If that had changed and we are onboarding the Proxy VM directly as a Proxy product and not as a SLES15SP4 product, then, we might not be using shared client tools, but still, depending the re-submission it could still affect proxy product.

File: terracumber-cli [hunk 1]
```diff
@@ -95,6 +94,10 @@ def parse_args():
     parser.add_argument('--bastion_ssh_key', help="""Bastion key to be use""", dest='bastion_ssh_key', default=None)
     parser.add_argument('--bastion_user', help="""Bastion user to be use""", dest='bastion_user', default='ec2-user')
     parser.add_argument('--bastion_hostname', help="""Bastion hostname to use""", dest='bastion_hostname', default=None)
+    parser.add_argument('--use-minion-cleaner', help='Activate or deactivate the minion cleaner',
```
Comment (srbarrios) [inline terracumber-cli]: In Terracumber code, we might want to abstract the concept of Salt Minions, that would make it more flexible if we decide to handle other kind of components.
Please consider my comments [1](https://github.com/uyuni-project/terracumber/pull/35#issuecomment-2277343672) and [2](https://github.com/uyuni-project/terracumber/pull/35#issuecomment-2277415555) together with this suggestion.
What about naming it `--use-tf-resource-cleaner`?

File: terracumber-cli [hunk 2]
```diff
@@ -95,6 +94,10 @@ def parse_args():
...
+                        action='store_true', default=False)
+    parser.add_argument('--minion-list', help='Space-separated list of minions',
```
Comment (srbarrios) [inline terracumber-cli]: For what suggested above, I would name it `--tf-resource-list` or similar.

**Comment by maximenoel8** [comment]
@srbarrios Build minion, terminal minion, monitoring-server and proxy were excluded from this first version because my first use case was related to have a limited deployment for BV using the minion list https://github.com/SUSE/susemanager-ci/blob/9b5de70b0c57ec609aac8c8192b90ed53b487696/jenkins_pipelines/environments/manager-4.3-qe-build-validation-NUE#L4-L23.

For me deleting those others modules should be an active action ( meaning declaring what we want to delete ) in comparison of the minion list that is more a passive action (removing the minions not in the list). 

In BV, we don't have an option to do that so far but we could have an active action for cleaning.

I added a new parameter call `--tf-resources-to-delete` that has a limited choices : ['proxy', 'monitoring-server', 'retail']

An example of use case :
```
python3.9 terracumber-cli --outputdir /home/maxime/github/sumaformMainCreator/ --tf ./SUSEManager-4.3-build-validation-NUE.tf --gitfolder /home/maxime/github/sumaformMainCreator/ --logfile /home/maxime/github/sumaformMainCreator/ --sumaform-backend libvirt --use-tf-resource-cleaner --tf-resources-to-delete proxy retail --tf-resources-to-keep sles15sp5-sshminion sles15sp6-sshminion sles15sp4-client rocky8-minion --runstep provision 
```

**Comment by maximenoel8** [comment]
I update the main description with the last suggestions from @srbarrios .

**Review by srbarrios — APPROVED** [review]

**Pull request merged**